charts/newrelic-infra-operator/templates/deployment.yaml

```diff
@@ -0,0 +1,87 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ template "newrelic-infra-operator.fullname" . }}
+  labels:
+    {{- include "newrelic-infra-operator.labels" . | nindent 4 }}
+spec:
+  replicas: {{ .Values.replicas }}
+  selector:
+    matchLabels:
+      {{- include "newrelic-infra-operator.appLabel" . | nindent 6 }}
+  template:
+    metadata:
+      {{- if .Values.podAnnotations }}
+      annotations:
+        {{- toYaml .Values.podAnnotations | nindent 8 }}
+      {{- end }}
+      labels:
+        {{- include "newrelic-infra-operator.labels" . | nindent 8 }}
+    spec:
+      serviceAccountName: {{ template "newrelic-infra-operator.serviceAccountName" . }}
+      {{- if .Values.podSecurityContext.enabled }}
+      securityContext:
+        runAsUser: {{ .Values.podSecurityContext.runAsUser }}
+        runAsGroup: {{ .Values.podSecurityContext.runAsGroup }}
+        fsGroup: {{ .Values.podSecurityContext.fsGroup }}
+      {{- end }}
+      {{- if .Values.image.pullSecrets }}
+      imagePullSecrets:
+        {{- toYaml .Values.image.pullSecrets | nindent 8 }}
+      {{- end }}
+      containers:
+      - name: {{ template "newrelic-infra-operator.name" . }}
+        image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
+        imagePullPolicy: "{{ .Values.image.pullPolicy }}"
+        env:
+        - name: CLUSTER_NAME
+          value: {{ include "newrelic-infra-operator.cluster" . }}
+        - name: RELEASE_NAME
+          value: {{ template "newrelic-infra-operator.fullname" . }}
+        - name: NRIA_LICENSE_KEY
+          valueFrom:
+            secretKeyRef:
+              {{- if (include "newrelic-infra-operator.licenseKey" .) }}
+              name: {{ template "newrelic-infra-operator.fullname" . }}-config
+              key: license
+              {{- else }}
+              name: {{ include "newrelic-infra-operator.customSecretName" . }}
```

Comment:
If both `customSecretName` and `licenseKey` are empty, we should throw an error while rendering.

Reply:
Previously, the approach that the charts followed were to have a helper checking for this, and surround every template with an `if` against that helper's output.

I'm not much of a fan of that approach but can't think about anything better right away, do you have any suggestion?

Reply:
You can use `required` in conditional in the template, perhaps that would work? I'm not sure exactly what is the issue here.

Reply:
I think it's a bit ugly since we're not using "one value or the other" in the same place.  What are your thoughts about:

```diff
diff --git a/charts/newrelic-infra-operator/templates/deployment.yaml b/charts/newrelic-infra-operator/templates/deployment.yaml
index 7c696ed..8f5f03a 100644
--- a/charts/newrelic-infra-operator/templates/deployment.yaml
+++ b/charts/newrelic-infra-operator/templates/deployment.yaml
@@ -44,9 +44,11 @@ spec:
               {{- if (include "newrelic-infra-operator.licenseKey" .) }}
               name: {{ template "newrelic-infra-operator.fullname" . }}-license
               key: license
-              {{- else }}
+              {{- else if include "newrelic-infra-operator.customSecretName" . }}
               name: {{ include "newrelic-infra-operator.customSecretName" . }}
               key: {{ include "newrelic-infra-operator.customSecretLicenseKey" . }}
+              {{- else }}
+              {{- nil | required "Cannot find License Key, either licenseKey or customSecretName must be defined" }}
               {{- end }}
         volumeMounts:
         - name: config
```

Reply:
Looks good to me. We could even put the whole license part into the template on it's own to make the `deployment.yaml` a bit simpler.